<div style="border:solid steelblue 3px; padding: 20px; border-radius: 10px">

**Привет, Андрей!** 👋

Меня зовут Алексей Манелюк, я буду проверять твой проект. В Яндекс Практикуме принято общаться на "ты" 😉, но если некомфортно - скажи и перейдем на "Вы".

Моя цель - не указать на ошибки, а поделиться опытом и помочь в освоении анализа данных.

В тетрадке ты увидишь мои комментарии, пожалуйста, не удаляй, не перемещай и не изменяй их. У комментариев могут быть приписки **V2**, **V3**, **V...**, которые соответствуют каждой версии проверки. 
    
Я буду использовать цветовую разметку:
<p>
<div class="alert-danger" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ❌</b>

Так выделены критические замечания. Без их отработки я не смогу принять проект. </div>

<div class="alert-warning" style="padding: 20px; border-radius: 10px">
<b>  Комментарий ревьюера ⚠️</b>

Так выделены небольшие замечания или полезные советы. Постарайся учесть их.</div>
<p>
<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Так выделены правильные решения.</div>

Предлагаю работать в диалоге: если что-то дорабатываешь или отвечаешь на мой комментарий - пиши об этом. Мне будет удобней, если ты выделишь свои комментарии:
<p>
<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Здесь исправил.</div>

Все это поможет мне быстрее принять твой проект.</div>

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Написать свой комментарий можно, скопировав этот блок в ячейку Markdown.</div>

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



<div class="alert-warning" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ⚠️</b>

Несмотря на то, что первая часть проекта проверяется автоматизрованно, дам по ней пару советов - они пригодятся тебе в следующих проектах. 

**Их отработка сейчас необязательна**.

📌 После названия, обязательно нужно добавлять бриф, цели исследования, описание данных и план работы.</div>

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert-warning" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ⚠️</b>

📌 Каждый раздел работы стоит завершать кратким выводом, о том что было сделано и какие результаты получены.

Как правильно оформить проект ты можешь прочитать в материалах курса:

`Полезные инструкции для учёбы - Оформление проекта - Рекомендации по выполнению проектов`
</div>

## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for x in data['income_type'].unique():
    data.loc[(data['income_type'] == x) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == x), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data_children = data.pivot_table(index = [ 'children'], values = 'debt', aggfunc = {'count','sum'}) # строим сводную таблицу
data_children.columns = ['total', 'debt'] #переименовываем столбцы
data_children['default_chance'] = data_children['debt']/data_children['total'] # добавляем столбец с оценкой шанса на дефолт
data_children['default_chance'] = data_children['default_chance'].map('{:.1%}'.format) # переводим значения в этом столбце в процентный формат
display(data_children)

#Рассчитаем соотношение между шансами на просрочку у тех, кто имееет имеет хотя бы одного ребенка к шансам на просрочку у тех, у кого нет детей

absolute_no_children = (data[data['children'] == 0]['debt'].sum())/(data[data['children'] == 0]['debt'].count())
absolute_children = (data[data['children'] != 0]['debt'].sum())/(data[data['children'] != 0]['debt'].count())
children_difference = absolute_children/absolute_no_children
children_difference_percent = (float("{:.2f}".format(((children_difference - 1)*100)))) # Не знаю, как более грамотно перевести в процентный формат с двумя знаками после запятой
print('Наличие ребенка повышает вероятность дефолта по кредиту на', children_difference_percent, '%')


total  debt default_chance
children                            
0         14091  1063           7.5%
1          4808   444           9.2%
2          2052   194           9.5%
3           330    27           8.2%
4            41     4           9.8%
5             9     0           0.0%

Наличие ребенка повышает вероятность дефолта по кредиту на 22.49 %


<div class="alert-warning" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ⚠️</b>
    
Это правильная сводная таблица для сравнения категорий. 👌

📌 Аналогичную сводную таблицу можно вывести, передав весь список агрегирующих функций: 👇</div>

In [23]:
# Код ревьюера
data.pivot_table(index = 'children', values = 'debt', aggfunc = ['count', 'sum', 'mean'])

count   sum      mean
           debt  debt      debt
children                       
0         14091  1063  0.075438
1          4808   444  0.092346
2          2052   194  0.094542
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Зависимость между группами с 0, 1, 2 детьми описал в общем выводе.</div>

**Вывод:** По приведенной таблице видна зависимость вероятности просрочки по кредиту от количества детей. Сам факт наличия ребенка повышает вероятность дефолта на 22.49 процента. Если сравнивать заемщиков с детьми, разница в количестве детей практически не влияет на вероятность дефолта по кредиту.

<div class="alert-success" style="border-radius: 10px; padding: 10px">
<b>Комментарий ревьюера V3 ✔️</b>

Я бы сказал, что если сравнить только самые многочисленные выборки с 0, 1 и 2 детьми, можно сказать, что чем больше у заемщика детей, тем выше риск допустить просрочку.

Даже незначительная разница в проценте должников будет важна для заказчика.</div>

<div class="alert-warning" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ⚠️</b>

Обрати внимание на то, что выборки несбалансированны. Поэтому выявлять зависимость стоит, сравнивая самые большие, сопоставимые по размеру категории заемщиков. Какую зависимость можно выявить, сравнив категории с 0, 1 и 2 детьми?</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
data_family = data.pivot_table(index='family_status', columns='debt', values='children', aggfunc='count')
data_family.columns = ['no_debt', 'debt']
data_family['total'] = data_family['debt'] + data_family['no_debt']
data_family['default_chance'] = data_family['debt'] / data_family['total']
data_family['default_chance'] = data_family['default_chance'].map('{:.1%}'.format)
display(data_family)

#Сгруппируем тех, кто не женат, и тех, кто находится в гражданском браке, в одну категорию. 
#Посчитаем разницу в вероятности дефолта для этой группы по отношению к остальным значениям выборки.

not_married = data[data['family_status'] != 'в разводе']    # ошибаюсь в синтаксисе, когда фильтрую с несолькими условиями, поэтому фильтрую последовательно
not_married = not_married[not_married['family_status'] != 'вдовец / вдова']
not_married = not_married[not_married['family_status'] != 'женат / замужем']
other_status = data[data['family_status'] != 'Не женат / не замужем']
other_status = other_status[other_status['family_status'] != 'гражданский брак']

not_married_default_chance = (not_married[not_married['debt'] == 1]).count()/not_married.count()
other_status_default_chance = (other_status[other_status['debt'] == 1]).count()/other_status.count()
percentage = not_married_default_chance['family_status']/other_status_default_chance['family_status']
percentage = (float("{:.2f}".format(((percentage - 1)*100)))) # Форматирую аналогично предыдущей задаче
print('Вероятность дефолта по кредиту для неженатых и проживающих в гражданском браке заемщиков по сравнению с остальными группами выше на', percentage, '%')


no_debt  debt  total default_chance
family_status                                             
Не женат / не замужем     2523   273   2796           9.8%
в разводе                 1105    84   1189           7.1%
вдовец / вдова             888    63    951           6.6%
гражданский брак          3749   385   4134           9.3%
женат / замужем          11334   927  12261           7.6%

Вероятность дефолта по кредиту для неженатых и проживающих в гражданском браке заемщиков по сравнению с остальными группами выше на 27.32 %


**Вывод:** По таблице видно, что две выдающиеся категории - не женатые /не замужем и гражданский брак. Сгруппировав эти две категории, выясним, что вероятность дефолта для этих двух категорий выше, чем у остальных на 27.32 процента.
Самой зажиточной категорией являются вдовцы/вдовы - вероятность неуплаты по кредиту ниже всех - всего  6.6% процента.

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Верно, если обобщить, можно сказать, что состоящие или состоявшие в браке заемщики реже допускают просрочки.</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [25]:
data_income = data.pivot_table(index='total_income_category', columns='debt', values='children', aggfunc='count')
data_income.columns = ['no_debt', 'debt']
data_income ['total'] = (data_income['debt'] + data_income['no_debt'])
data_income['default_chance'] = data_income['debt'] / data_income ['total']
data_income['default_chance'] = data_income['default_chance'].map('{:.1%}'.format)

display(data_income)

# Промежуточный вывод - данное распределение на категории нерепрезентативно. Большинство заемщиков попадают в категории С и В 
# с доходом от 50 до 200 тысяч рублей и от 200 тысяч до 1 миллтона рублей соответственно
# Для более подробного анализа необходимо ввести больше деления по группам

def super_categorize_income(income):
    try:
        if 0 <= income <= 50000:
            return 'E'
        elif 50001 <= income <= 80000:
            return 'D'
        elif 80001 <= income <= 125000:
            return 'C'
        elif 125001 <= income <= 180000:
            return 'B'
        elif 180001 <= income <= 250000:
            return 'A2'
        elif 250001 <= income <= 500000:
            return 'A1'
        elif 500001 <= income:
            return 'A'
    except:
        pass
    
    
data['super_income_category'] = data['total_income'].apply(super_categorize_income)
print()
#Проверим сколько человек в каждой группе
print(data.groupby('super_income_category')['super_income_category'].count())
#и построим новую сводную таблицу

data_income_sorted = data.pivot_table(index='super_income_category', columns='debt', values='children', aggfunc='count')
data_income_sorted.columns = ['no_debt', 'debt']
data_income_sorted ['total'] = (data_income_sorted['debt'] + data_income_sorted['no_debt'])
data_income_sorted['default_chance'] = data_income_sorted['debt'] / data_income_sorted ['total']
data_income_sorted['default_chance'] = data_income_sorted['default_chance'].map('{:.1%}'.format)

display(data_income_sorted)


    

no_debt  debt  total default_chance
total_income_category                                     
A                           23     2     25           8.0%
B                         4660   354   5014           7.1%
C                        14568  1353  15921           8.5%
D                          328    21    349           6.0%
E                           20     2     22           9.1%


super_income_category
A      222
A1    2575
A2    3592
B     7226
C     5451
D     1894
E      371
Name: super_income_category, dtype: int64


no_debt  debt  total default_chance
super_income_category                                     
A                          208    14    222           6.3%
A1                        2395   180   2575           7.0%
A2                        3310   282   3592           7.9%
B                         6612   614   7226           8.5%
C                         4983   468   5451           8.6%
D                         1743   151   1894           8.0%
E                          348    23    371           6.2%

**Вывод:** После применения нового метода сортировки очевидна тенденция к уменьшению шансов дефолта по мере роста дохода заемщика.

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Здорово, что ты провел дополнительную категоризацию заемщиков.
    
Все верно. Если сравнить самые большие выборки B и C из первой таблицы, такая зависимость видна.</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [26]:
data_goal = data.pivot_table(index='purpose_category', columns='debt', values='children', aggfunc='count')
data_goal.columns = ['no_debt', 'debt']
data_goal['total'] = (data_goal['debt'] + data_goal['no_debt'])
data_goal['default_chance'] = data_goal['debt'] / data_goal['total']
data_goal['default_chance'] = data_goal['default_chance'].map('{:.1%}'.format)
display(data_goal)

no_debt  debt  total default_chance
purpose_category                                             
операции с автомобилем       3879   400   4279           9.3%
операции с недвижимостью     9971   780  10751           7.3%
получение образования        3619   369   3988           9.3%
проведение свадьбы           2130   183   2313           7.9%

**Вывод:** Самые рискованные категории по целям кредита - на операции с автомобилем и на образование. Меньше всего просрочек - на кредиты на операции с недвижимостью.

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Всё правильно. 👌</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Человеческий фактор, технический сбой, повреждения носителя, ограничение доступа, неправильная организация хранения данных и другие

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Предположения о возникновении пропусков правильные. </div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медианные значения меньше всего искажают результат исследования, т.к. наименее подвержены влиянию экстремальных значений в датасете.

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Совершенно верно, медиана менее чувствительна к выбросам.</div>

### Шаг 4: общий вывод.

**Состояние датасета**

Исследование выявило несколько ошибок в данных.

Во-первых, сразу заметны экстремальные значения трудового стажа у безработных и пенсионеров. Учитывая особенности данных категорий, вероятно, целесообразно было либо изменить сбор данных таким образом, чтобы исключить ошибки, либо полностью отказаться от сбора данных в данной категории (т.к. трудовой стаж не является критическим показателем при принятии решения о выдаче кредита ни для пенсионеров, ни для безработных).

Во-вторых, считаю целесообразным категоризировать цель получения кредита на первоначальном этапе сбора данных. Предпочтительно, чтобы сам заемщик выбирал из предложенных категорий цели кредита на этапе подачи заявки, а не вписывал цель от руки. Предлагаю следующий набор категорий: "операции с авто", операции с недвижимостью", "получение образования", "проведение свадьбы", "другое".

В-третьих, не считаю сообразным заполнять пропуска в позициях "days_employed" и "total_income", как того требует задание 5.
Скорее всего, пропуски обусловлены тем, что у этих людей нет работы и трудового стажа. Соответственно, заполнив эти позиции медианой, как предписано в задании, мы сильно исказим данные.
Можно предположить, что наличие пропусков в данных позициях как-то связано с экстремальными значениями трудового стажа у песнионеров и безработных, однако, это требует дополнительного исследования, которое не предусмотрено заданиями данного проекта.

Категоризация доходов заемщиков (как требуется в задании) недостаточно репрезентативно разделяет их на группы. Для более глубокой оценки данного фактора, необходимо разделение на большее количество групп (см. задание 3.3)

**Выявленные закономерности**

Наличие у потенциального заемщика ребенка повышает шансы дефолта по кредиту более чем на 22 процента (если сравнивать с заемщиками с детьми).
Самая многочисленная группа заемщиков - заемщики без детей, далее размеры групп уменьшаются по мере роста количества детей.
Группы с 3 и более детей не достаточно репрезентативны в виду своей малочисленности.
Если сравнивать группы с 0, 1, 2 детьми риск дефолта растет с увеличением числа детей.

Вероятность дефолта по кредиту для неженатых и проживающих в гражданском браке заемщиков по сравнению с остальными группами выше на 27.32 %, иначе говоря состоящие или состоявшие в браке заемщики реже допускают просрочки.

Уровень дохода заемщика прямо пропорционален его надежности. Чем выше доход - тем надежнее является заемщик.

Если оценивать по цели кредита - меньше всего рисков невозврата у тех, кто берет заем на операции с недвижимостью. Далее - в порядке возрастания риска - кредит на проведение свадьбы. Наконец, наибольший риск невозврата у кредитов на операции с авто и на получение образования.



<div class="alert-success" style="border-radius: 10px; padding: 10px">
<b>Комментарий ревьюера V3 ✔️</b>

Теперь у тебя получился очень подробный общий вывод, отражающий все основные результаты исследования. 👌</div>

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Написал новый вывод с учетом комментариев. Версия в браузере не сохранилась, сейчас попробовал сделать оффлайн, надеюсь в этот раз все будет ок.</div>

<div class="alert-danger" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ❌</b>

Общий вывод в целом верный, однако стоит его расширить. Нужно отразить в нем основные результаты по всем шагам работы. Начать с абзаца с описанием входных данных и что сделали на этапе предобработки. 

Далее описать все выявленные зависимости просрочки от различных факторов в общем виде. 

Закончить вывод можно рекомендациями банку по улучшению сбора данных.

Зачастую, заказчик читает только общий вывод, поэтому он должен быть максимально информативным.

</div>

<div style="border:solid steelblue 3px; padding: 20px; border-radius: 10px">
    
<b>Общий комментарий ревьюера:</b>

Я вижу, что ты потратил много времени и усилий, чтобы выполнить этот проект на высоком уровне. Но нужно кое-что исправить, чтобы я его принял.
    
У меня только одно критическое замечание - касательно общего вывода. Его нужно раскрыть поподробнее.
    
Также, посмотри желтые комментарии во второй части работы.
    
Буду ждать твой доработанный проект. ⌛️</div>

<div style="border:solid steelblue 3px; padding: 20px; border-radius: 10px">
    
<b>Общий комментарий ревьюера V2</b>

К сожалению, не увидел реакции на свои комментарии. Возможно что-то пошло не так и твои правки не сохранились.
    
</div>

<div style="border:solid steelblue 3px; padding: 20px; border-radius: 10px">
    
<b>Общий комментарий ревьюера V3</b>

Спасибо, что доработал проект! Критических замечаний больше нет, получилась очень хорошая работа.
    
Успехов в дальнейшем обучении! 🏆
    
</div>